tests/std/tests/VSO_0157762_feature_test_macros/test.cpp

```diff
@@ -5,465 +5,431 @@
 
 int main() {} // COMPILE-ONLY
 
-// COMPILER FEATURE-TEST MACROS
+// ATTRIBUTE FEATURE-TEST MACROS.
 
-#ifdef _MSVC_LANG
-#if _MSVC_LANG > 201703L
-#define CXX17_MODE 1
-#define CXX20_MODE 1
-#elif _MSVC_LANG > 201402L
-#define CXX17_MODE 1
-#define CXX20_MODE 0
+#ifdef __has_cpp_attribute
+// Good.
 #else
-#define CXX17_MODE 0
-#define CXX20_MODE 0
+#error Expected __has_cpp_attribute to be defined.
 #endif
+#if defined(__has_cpp_attribute)
+// Good.
 #else
-#if __cplusplus > 201703L
-#define CXX17_MODE 1
-#define CXX20_MODE 1
-#elif __cplusplus > 201402L
-#define CXX17_MODE 1
-#define CXX20_MODE 0
-#else
-#define CXX17_MODE 0
-#define CXX20_MODE 0
+#error Expected __has_cpp_attribute to be defined.
 #endif
+#ifndef __has_cpp_attribute
+#error Expected __has_cpp_attribute to be defined.
 #endif
 
+#if __has_cpp_attribute(carries_dependency) != 200809L
+#error Expected has_cpp_attribute(carries_dependency) to equal 200809L.
+#endif
 
-// Always defined to varying values (C++14 versus C++17-and-newer mode).
+#if __has_cpp_attribute(deprecated) != 201309L
+#error Expected has_cpp_attribute(deprecated) to equal 201309L.
+#endif
 
-#ifndef __cpp_constexpr
-#error BOOM
-#elif CXX17_MODE
-#if __cpp_constexpr != 201603L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_constexpr == 201603L);
+#if __has_cpp_attribute(noreturn) != 200809L
+#error Expected has_cpp_attribute(noreturn) to equal 200809L.
+#endif
+
+#if defined(__clang__) || defined(__EDG__) // clang and EDG don't yet implement P1771R1
+#if __has_cpp_attribute(nodiscard) != 201603L

```

Comment:
"nodiscard" < "noreturn" [My expectation is that within the attribute/core/library sections the tests are ordered lexicographically by the name of the attribute/feature-test macro to reflect the order in which the standard defines them.]

Reply:
The tests were also originally ordered by standard mode which is how I have kept them. Should that be kept at all or just completely lexicographically by test name?

Reply:
The issue stated that lexicographically was the proposed solution since there doesn't seem to be much value keeping them ordered by standard mode.
I would actually propose that we sort them lexicographically now but in the future we move to just appending new tests to the end of the file so in the future they are chronologically ordered by appearance in our library or in one of the supported compilers.

Reply:
Great! Because that was already what I was working on and now have just pushed.

Reply:
After talking to @StephanTLavavej I take back my appending comment. Let's keep this file sorted in the future as well.